tasmota/tasmota_xdrv_driver/xdrv_92_esp32_gnss.ino

```diff
@@ -0,0 +1,876 @@
+  the Free Software Foundation, either version 3 of the License, or
+  any later version.
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifdef ESP32
+#ifdef USE_GNSS
+
+/*********************************************************************/
+/*                          INCLUDES                                 */
+/*********************************************************************/
+#include <queue>
+#include <array>
+#include <cstring>
+#include <TinyGPSPlus.h>
+#include <GNSSParser.h>
+#include <vector>
+#include <algorithm>
+
+#ifdef USE_AUX_WEBSERVER
+#include <ESPAsyncWebServer.h>
+#endif
+
+#ifdef USE_NTRIP
+extern void ProcessRTCMMessage(const uint8_t *data, size_t length);
+#endif
+
+/*********************************************************************/
+/*                          DEFINES                                  */
+/*********************************************************************/
+#define XDRV_92 92
+#define D_CMND_GNSS "GNSS"
+#define RTC_UPDATE_INTERVAL (30 * 60) // Every 30 minutes
+#define INVALIDATE_AFTER 30           // 30 seconds
+#define SERIAL_INPUT_BUFFER_SIZE 2048
+#define SERIAL_STREAM_BUFFER_SIZE 4096
+
+#ifdef USE_AUX_WEBSERVER
+#define MAX_STREAMING_CLIENTS 10 // Maximum number of simultaneous streaming clients
+#define CLIENT_TIMEOUT 30000     // 30 seconds timeout for streaming clients
+#define AUX_WEBSERVER_PORT  8080
+#endif
+
+/*********************************************************************/
+/*                          STRUCTURES                               */
+/*********************************************************************/
+struct NMEACounters
+{
+  uint32_t gga;   // Global Positioning System Fix Data
+  uint32_t rmc;   // Recommended Minimum Navigation Information
+  uint32_t gll;   // Geographic Position - Latitude/Longitude
+  uint32_t gsa;   // GNSS DOP and Active Satellites
+  uint32_t gsv;   // GNSS Satellites in View
+  uint32_t vtg;   // Course Over Ground and Ground Speed
+  uint32_t other; // Any other NMEA sentences
+};
+
+struct SerialConfigMap
+{
+  const char *name;
+  uint32_t config;
+};
+
+typedef struct
+{
+  uint32_t crc32;
+  uint32_t version;
+  uint32_t baudrate;
+  uint32_t serial_config;
+} gnss_settings_t;
+
+#ifdef USE_AUX_WEBSERVER
+struct StreamClient
+{
+  AsyncClient *client;
+  uint32_t lastDataTime;
+  bool isActive;
+};
+#endif
+
+/*********************************************************************/
+/*                   CLI COMMANDS                                    */
+/*********************************************************************/
+
+void CmndGNSSBaudrate(void);
```

Comment:
Due to the magic of Arduino, you are not strictly required to make these early declarations. But they don't hurt either.